pkg/kubelet/kubelet_getters.go

```diff
@@ -237,14 +235,8 @@ func (kl *Kubelet) GetNode() (*v1.Node, error) {
 	if kl.kubeClient == nil {
 		return kl.initialNode(context.TODO())
 	}
-	// if we have a valid kube client, we wait for initial lister to sync
 	if !kl.nodeHasSynced() {
-		err := wait.PollImmediate(time.Second, maxWaitForAPIServerSync, func() (bool, error) {
-			return kl.nodeHasSynced(), nil
-		})
-		if err != nil {
-			return nil, fmt.Errorf("nodes have not yet been read at least once, cannot construct node object")
-		}
+		return nil, fmt.Errorf("nodes have not yet been read at least once, cannot construct node object")
```

Comment:
is it possible that returning the error can cause some regressions in future at places like below?
https://github.com/kubernetes/kubernetes/blob/c78f67a005977bf1159b538a33c2ee3bf5ca24da/pkg/kubelet/kubelet.go#L1668-L1670

if ` fastStatusUpdateOnce` is the only cause of concern here, a new `getNode` method can be introduced for it?
https://github.com/kubernetes/kubernetes/blob/c78f67a005977bf1159b538a33c2ee3bf5ca24da/pkg/kubelet/kubelet.go#L2268-L2274

Reply:
all places that need the Node object are already separate threads / go routines and they retry until they get a valid node object. i don't see a case where a node object fetch would not be retried in the future.

Reply:
i had debated waiting for the informer to sync much earlier in the kubelet startup path before any GetNode invocation occurred.  for example, i had debated putting the blocking wait in kubelet.Run() before all other modules were initialized.

Reply:
i tried figuring out how to do that yesterday, but i think its a bit complicated because the static pod loop has to start (for the kubeapiserver static pod to start, and for the sync to complete), but in the meantime everything else has to wait for this before calling GetNode().

currently a number of different routines start all at once, some of which consume GetNode().

Reply:
it also seems to me requires a bigger refactor that is not ideal to backport.

Reply:
@neolit123 thanks for confirming what may have been what i discovered and had forgot when fixing the described issue.

Reply:
btw, this message might need adjustment:
https://github.com/kubernetes/kubernetes/blob/c78f67a005977bf1159b538a33c2ee3bf5ca24da/pkg/kubelet/kubelet.go#L1668-L1670

since, `err != nil` for `GetNode` does not necessarily mean "node is deleted".